New Feature: Bulk Shifting Activities

shreyasyellenki opened the pull request:
Added a new feature that allows the user to shift `Activity Directives` either directions. Shift in this instance refers to the desired time difference in absolute start time. This action is done through the `Activity Directives Table`, where a modal allows the user to decide direction and the desired shift. Users are warned of invalid shift (including negative values as the behavior can instead be replicated through packing the opposite direction). This was originally a sub-branch from the  `feature/pack-activities-offset` so it holds commit history from those branches, however all code relating to packing has been removed in latest commit to make code review easier.

shreyasyellenki committed:
Beginning of PackLeftAction

Working towards creating the packLeft function, and helper methods to parse Stings of format "HH:mm:ss.SSSSSS" into microseconds. There seems to be some issues with the hasura query and the format of the duration string that needs to be reworked.

shreyasyellenki committed:
Pack Left

Have an early working implementation of packLeft, it currently works on the framework of activities can only be packed if they have the same anchor, "null" is a valid anchor.

shreyasyellenki committed:
Revamping packLeft

Working on a new implementation of packLeft that allows packing of various anchored activities but follows a specific set of defined rules discussed in a meeting. Will add documentation later to explain the rules

shreyasyellenki committed:
Merge remote-tracking branch 'upstream/develop' into feature/pack-actions

shreyasyellenki committed:
Reworked Pack Left

Have a new working implementation of packLeft that covers the edge cases of anchors. Some refactoring still needs to be done and then replication for packRight, also realized that initialStartTime is only necessary to keep track of for the first activity and not a map to all activities.

shreyasyellenki committed:
Reset Pack Activities

Revamped packLeft with the new guidelines involving anchor activities discussed in the Aerie-External Sim Shallow Integration Working Group Meeting. This new approach is in a function with the temporary name packActivityDirectivesInPlanRevamp. I have some ideas to use the same method for left and right with an enum switch between the two but that is something I am still testing, hence the reason it is in a temporary named funciton for now.

shreyasyellenki committed:
fixed packRight and refactored the appropriate packLeft

Have a working implementation of packRight and updated condition check for if the activity being packed was an anchor to other activities (it initally didn't do anything because it was the wrong conditional). Removed the old packLeftActivityDirectivesInPlan implementation and refactored the "revamp" method to the appropriate name.

shreyasyellenki committed:
Introduced a Less cluttered version of packing Activities

Combined both implementations of packing Activities in a single method using a const union type for direction. Will be talking with Aerie-UI team if this is fine for the development standards, for now not deleting the code for the separate options as well.

shreyasyellenki committed:
Started building PackActivitiesOffsetSelect

Built a dialog for packing activity directive with offset, the cancel functionality works but the pack button still needs to be updated. First need to make some refactoring changes to pack actions.

shreyasyellenki committed:
Code cleanup for packActivityDirectivesInPlan

Deleted the separate methods for packing activities to left and right, and kept packActivityDirectivesInPlan which handles both cases. Removed unnecessary code and used aerie-ui stores instead to achieve same functionality.

shreyasyellenki committed:
Merge branch 'feature/pack-actions' into feature/pack-actions-offset
Refactored packActivityDirectivesInPlan so it is neater, removed unnecessary code.

shreyasyellenki committed:
Added offset functionality

Instantiated an offset field (in microseconds) to packing Activities. This will play a more vital role in the development of the pack-actions-offset feature that is also being developed. When the regular 'Pack Left' and 'Pack Right' options are selected the offset field is 0.

shreyasyellenki committed:
Merge branch 'feature/pack-actions' into feature/pack-actions-offset
Merging the offset parameter introduced to packActivityDirectivesInPlan method.

shreyasyellenki committed:
Working Implementation of Pack with Offset

Have a working implementation of packing activities with offset. A dialog pops up that allows the user to pick a direction and offset (default to 0).

shreyasyellenki committed:
Updated design for PackActivitiesOffsetSelect

Changed the offset string box to the appropriate CSS for error, don't allow negative offsets to be entered. The user cannot pack if the Offset string input is invalid.

shreyasyellenki committed:
Working Implementation of Bulk Shifting Activities

Have a working implementation of bulk shifting activities. The user receives a dialog allowing them to pick a direction and duration (default to 0) to shift. The user can only shift if they enter a valid duration.

shreyasyellenki committed:
Changed location of where database gets updated

Felt it would be more appropriate if the database was updated from ActivityDirectivesTable and packActivityDirectivesInPlan just returns the array of updated activities instead. This also simplifies when writing unit tests

shreyasyellenki committed:
Merge branch 'feature/pack-actions' into feature/pack-actions-offset

shreyasyellenki committed:
Update Packing with Offset to match new packing framework

Updated packActivityWithOffset so it the updates to the database are done in the ActivityDirectivesTable, also simplified updating an array of ActivityDirectives into one method

shreyasyellenki committed:
Merge branch 'feature/pack-actions-offset' into feature/bulk-shift

shreyasyellenki committed:
Merged with packActivitiesWithOffset

Merged the two branches and staying with the universal idea that updating the database is done outside of the activities.ts and rather inside ActivityDirectivesTable. Some more changes need to be made for bulk shifting activities to make that possible.

shreyasyellenki committed:
Fixed directional mishap for packLeft

During refactoring to move updating the database outside of the utility function itself, packLeft was accidentally affected so it would move right. That has now been fixed.

shreyasyellenki committed:
refactor: modal design

Change the current display modal for bulk shift to utilize the existing Aerie modal in svelte

shreyasyellenki committed:
refactor: bulk modal design

Change design aspects of bulk shift modal, will get it reviewed by Aerie team members

shreyasyellenki committed:
refactor: bulk shift workflow

Change bulkShiftActivityDirectivesInPlan so it only returns the updated activities. The Database update is now in ActivityDirectiveTable.svelte to match the packing functions. Minor design changes to BulkdShiftActivitiesModal, now has a header and buttton and input height are the same.

shreyasyellenki committed:
refactor: remove async, test: add bulkShiftActivityDirectivesInPlan

Since, the database update happens outside bulkShiftActivityDirectivesInPlan it no longer has to be an async function. Added unit for bulk shift for both left and right direction.

shreyasyellenki committed:
Refactor: remove pack Activity Directives code

Delete all code related to pack Activities to make code review easier

shreyasyellenki committed:
Merge remote-tracking branch 'upstream/develop' into feature/bulk-shift

shreyasyellenki committed:
refactor: fix change made to test

Revert accidental change made to a addAbsoluteTimeToRevision unit test

duranb commented on src/components/modals/BulkShiftActivitiesModal.svelte:
```diff
@@ -0,0 +1,167 @@
+  import { getTarget } from '../../utilities/generic';
+  import { convertDurationStringToInterval } from '../../utilities/time';
+  import { tooltip } from '../../utilities/tooltip';
+  // import Input from '../form/Input.svelte';
```
Please remove any commented out code.

duranb commented on src/utilities/time.ts:
```diff
@@ -928,3 +928,24 @@ export function formatMS(ms: number | null): string {
+  us %= 1_000_000;
+  const micro = us;
+
+  const pad = (n: number, len: number) => n.toString().padStart(len, '0');
```
In general we avoid creating nested functions unless we absolutely need to scope it to the outer function. In this case, it could be a useful function for other parts of the codebase, so I would rather that this be added to the `generic.ts` file as `padNumber()`(or something similar) with some tests.

shreyasyellenki replied:
This has been moved to `time.ts` instead (ran this through Aaron Plave) due to `generic.ts` using `browser` causing some issues. Unit tests have been added to `time.test.ts`.

dandelany replied:
This eventually ended up in `text.ts`, with tests in `text.test.ts`

duranb commented on src/utilities/activities.ts:
```diff
@@ -282,6 +284,34 @@ export async function getActivityDirectivesToPaste(
+  const updateActivities = activities.filter(a => a.anchor_id === null || !selectedIds.has(a.anchor_id));
+
+  for (const activity of updateActivities) {
+    activity.start_offset = usToOffset(
```
Please avoid mutating objects directly as mutations are difficult to trace when debugging. I suggest something like returning the output of `updateActivities.map()` and cloning each activity via object spreading or something similar in order to set the `start_offset`.

duranb commented on src/utilities/time.ts:
```diff
@@ -928,3 +928,24 @@ export function formatMS(ms: number | null): string {
+ * Converts microseconds to a string of the form "HH:mm:ss.SSSSSS".
+ * Example: 5025678901 => "01:23:45.678901"
+ */
+export function usToOffset(us: number): string {
```
Do you mind writing some unit tests for this function, please?

duranb commented on src/components/ui/DataGrid/BulkActionDataGrid.svelte:
```diff
@@ -39,6 +41,8 @@
   export let selectedItemIds: RowId[] = [];
   export let showContextMenu: boolean = true;
   export let showCopyMenu: boolean = false;
+  export let showBulkShiftDialog = false;
```
The `BulkActionDataGrid` component is a generic component. The concept of "shifting" is very specific to activity directives. The component does expose a slot `context-menu` that you can use to insert the menu in to. The `ActivityDirectivesTable.svelte` file is already utilizing this so you can add to it there instead.

shreyasyellenki replied:
Just to clarify are you asking to move everything related to the display of the modal to the `ActivityDirectivesTable`?

duranb replied:
Yes, please!

duranb commented on src/components/modals/BulkShiftActivitiesModal.svelte:
```diff
@@ -0,0 +1,167 @@
+          <label
+            class="label"
+            use:tooltip={{ content: 'The duration of how much the activities should be shifted by', placement: 'top' }}
+            for="start-offset"
```
Should this be "gap-offset" instead?

duranb commented on src/components/modals/BulkShiftActivitiesModal.svelte:
```diff
@@ -0,0 +1,167 @@
+              >Left</button
+            >
+            <button class="toggle {direction === 'Right' ? 'active' : ''}" on:click={() => setDirection('Right')}
+              >Right</button
```
Same as above here

duranb commented on src/components/modals/BulkShiftActivitiesModal.svelte:
```diff
@@ -0,0 +1,167 @@
+          <div class="label">Direction</div>
+          <div class="toggle-group">
+            <button class="toggle {direction === 'Left' ? 'active' : ''}" on:click={() => setDirection('Left')}
+              >Left</button
```
Do you mind formatting this so that the `>` are on the same line as the rest of the component code?
i.e.
```
...setDirection('Left')}>
Left
</button>
```
instead of 
```
>Left</button
```

duranb commented on src/components/modals/BulkShiftActivitiesModal.svelte:
```diff
@@ -0,0 +1,167 @@
+  import ModalHeader from '../modals/ModalHeader.svelte';
+  // import ModalHeader from '../modals/ModalHeader.svelte';
+
+  const dispatch = createEventDispatcher();
```
`createEventDispatcher` accepts generic types. Do you mind typing the events that this component dispatches? You can use the [ActivityDirectivesTable](https://github.com/NASA-AMMOS/aerie-ui/blob/develop/src/components/activity/ActivityDirectivesTable.svelte#L34) as reference

duranb commented on src/components/modals/BulkShiftActivitiesModal.svelte:
```diff
@@ -0,0 +1,167 @@
+      <div class="button-container">
+        <button class="st-button secondary" on:click={cancel}>Cancel</button>
+        <button class="st-button" on:click={shift} disabled={!!shiftOffsetError}>Shift</button>
+      </div></ModalFooter
```
Similar formatting issue here as above in regards to the `>`

shreyasyellenki committed:
Style: Update for PR

Change bulkShiftActivityDirectivesInPlan so it is no longer in place. Make style changes to the modal as well as unit tests for new utility functions.

shreyasyellenki committed:
Refactor: Move BulkShift Modal display

Migrate the display of the bulk shift modal into ActivityDirectivesTable rather than BulkActionDataGrid.

shreyasyellenki committed:
Refactor: Extra slot in BulkActionDataGrid

Move bulk shift option display into extra slot of BulkActionDataGrid. Did this to match PR#1720

AaronPlave commented on src/components/activity/ActivityDirectivesTable.svelte:
```diff
@@ -210,6 +274,16 @@
     {/if}
+  </svelte:fragment>
+
+  <svelte:fragment slot="context-menu-alt">
```
Is there a reason to have this `context-menu-alt` slot instead of just moving the bulk shift menu into the `context-menu` slot above?

duranb replied:
Oh true. My initial comment on not moving the slot in the first place might have misled me into thinking to just add a new slot. I agree with @AaronPlave, if the new context menu items can just be added to the slot above, that would be preferred.

dandelany replied:
per discussion on other PR, we decided to keep this additional slot and rename to 'context-menu-bottom' for clarity

AaronPlave commented on src/components/modals/BulkShiftActivitiesModal.svelte:
Can't seem to use escape to close this modal

AaronPlave commented on src/components/activity/ActivityDirectivesTable.svelte:
Looks like there is now an extra separator in the menu?
<img width="267" height="249" alt="image" src="https://github.com/user-attachments/assets/7c220fdc-eafd-440f-a913-2a598b1f86f9" />

AaronPlave commented on src/components/modals/BulkShiftActivitiesModal.svelte:
```diff
@@ -0,0 +1,172 @@
+        <div class="row">
+          <div class="label">Direction</div>
+          <div class="toggle-group">
+            <button class="toggle {direction === 'Left' ? 'active' : ''}" on:click={() => setDirection('Left')}>
```
Could you use the `RadioButtons` component here?

dandelany replied:
not doing this for now - may change later

AaronPlave commented on .gitignore:
```diff
@@ -15,3 +15,4 @@ unit-test-results
 /playwright-report/
 /blob-report/
 /playwright/.cache/
+.env
```
Is this `.env` ignore needed?

duranb replied:
Good point. We explicitly don't want to ignore .env changes.

dandelany replied:
Fixed

AaronPlave commented on src/utilities/activities.ts:
```diff
@@ -282,6 +284,42 @@ export async function getActivityDirectivesToPaste(
 }
 
+export function findTypes(type: string, activityTypes: ActivityType[]): ActivityType | undefined {
+  // const activityTypes = await activityTypesPromise;
```
Remove unused code

AaronPlave commented on src/components/activity/ActivityDirectivesTable.svelte:
```diff
@@ -171,6 +183,58 @@
+  async function updateActivities(updatedActivities: ActivityDirective[] | null) {
+    if (plan != null && Array.isArray(updatedActivities)) {
+      for (const activity of updatedActivities) {
+        const activityType = findTypes(activity.type, get(planModelActivityTypes) ?? []);
```
Could this just be `const activityType = activityTypes.find(t => t.type === activity.type)`? Wouldn't need to helper function. Would also be good to move the `get(planModelActivityTypes) ?? []` to a variable outside of the loop.

AaronPlave commented on src/components/activity/ActivityDirectivesTable.svelte:
```diff
@@ -171,6 +183,58 @@
   }
 
+  async function updateActivities(updatedActivities: ActivityDirective[] | null) {
+    if (plan != null && Array.isArray(updatedActivities)) {
```
Can you use `!==` here?

AaronPlave commented on src/components/activity/ActivityDirectivesTable.svelte:
```diff
@@ -219,3 +293,7 @@
   </svelte:fragment>
 </BulkActionDataGrid>
+
+{#if showBulkShiftDialog}
```
Our pattern for modals is a bit boilerplate-y but what we do is run everything through the `effects.ts` file and add modals to the `modals.ts` file so that these functions can be called from anywhere and the modals don't need to be mounted in any specific component.

AaronPlave replied:
Would also let us more easily call `shift` on a directive from the activity timeline.

dandelany replied:
Updated to use `effects` + standard modals pattern, reviewed with @duranb

duranb commented on .env:
```diff
@@ -8,5 +8,12 @@ PUBLIC_HASURA_WEB_SOCKET_URL=ws://localhost:8080/v1/graphql
 PUBLIC_AUTH_SSO_ENABLED=false
 PUBLIC_TIME_PLUGIN_ENABLED=false
 PUBLIC_COMMAND_EXPANSION_MODE=typescript
+
```
Please revert these changes to .env. Any local overrides can be done by creating an `.env.local` file

duranb commented on src/components/activity/ActivityDirectivesTable.svelte:
```diff
@@ -21,6 +29,8 @@
 
   export let activityDirectives: ActivityDirective[] | null = null;
   export let activityDirectiveErrorRollupsMap: Record<ActivityDirectiveId, ActivityErrorRollup> | undefined = undefined;
+  export let showBulkShiftDialog = false;
```
Do we need to keep these props? I would say that they were only necessary for when they were in BulkActionDataGrid, but ActivityDirectivesTable probably would never need this toggled.

duranb commented on src/components/ui/DataGrid/BulkActionDataGrid.svelte:
```diff
@@ -214,6 +214,9 @@
         </div>
       {/if}
+
+      <slot name="context-menu-alt" />
```
If the comment about the context-menu results in combining the two context menu items into one group, then this slot can be removed.

dandelany replied:
changed to context-menu-bottom

duranb commented on src/components/ui/DataGrid/SingleActionDataGrid.svelte:
```diff
@@ -184,6 +184,8 @@
       </div>
     {/if}
+
+    <slot name="context-menu-alt" />
```
If the new slot is removed from BulkActionDataGrid, then please remove from here as well.

duranb commented on src/utilities/activities.ts:
```diff
@@ -282,6 +284,42 @@ export async function getActivityDirectivesToPaste(
+      return activityTypes[idx];
+    }
+  }
+  console.warn(`Activity type ${type} not found in activity types`);
```
Also remove this warn.

dandelany committed:
revert changes to .env and remove from gitignore

dandelany committed:
merge develop into feature/bulk-shift and resolve conflicts

dandelany committed:
run linter, small changes from review, remove unused findTypes util

dandelany committed:
fix context menu slots

dandelany committed:
Merge branch 'develop' into feature/bulk-shift

dandelany committed:
move 'pack activities' logic and modal creation to effects

dandelany committed:
fix linter

dandelany committed:
fix linter

dandelany committed:
refactor shift activities to effects and use consistent modal pattern

dandelany committed:
await results of async pack/shift activities

dandelany committed:
standardize button styles in pack/shift modals

dandelany commented:
🎉 thanks all, merging
